far/string_utils.hpp

```diff
@@ -180,6 +180,7 @@ class icase_searcher final: public i_searcher
 	exact_searcher m_Searcher;
 };
 
+template <typename Searcher>
```

Comment:
I'd prefer to not templatize the whole thing: fuzzy_searcher methods don't depend on the exact type of the underlying searcher and it would only duplicate the code. Maybe something like
<details>

```diff
diff --git a/far/string_utils.cpp b/far/string_utils.cpp
index 7a19e560c..b0e95a8ab 100644
--- a/far/string_utils.cpp
+++ b/far/string_utils.cpp
@@ -319,21 +319,19 @@ static void normalize_for_search(string_view const Str, string& Result, string&
 	Result.resize(End - Zip.begin());
 }
 
-static string normalize_for_search(string_view const Str, string& Intermediate, std::vector<WORD>& Types)
+fuzzy_searcher_impl::fuzzy_searcher_impl(i_searcher const* const Searcher):
+	m_Searcher(Searcher)
 {
-	string Result;
-	normalize_for_search(Str, Result, Intermediate, Types);
-	return Result;
 }
 
-template <typename Searcher>
-fuzzy_searcher<Searcher>::fuzzy_searcher(string_view const Needle, bool const CanReverse):
-	m_Searcher(normalize_for_search(Needle, m_Intermediate, m_Types), CanReverse)
+string fuzzy_searcher_impl::needle(string_view Needle) const
 {
+	string Result;
+	normalize_for_search(Needle, Result, m_Intermediate, m_Types);
+	return Result;
 }
 
-template <typename Searcher>
-std::optional<std::pair<size_t, size_t>> fuzzy_searcher<Searcher>::find_in(string_view const Haystack, bool const Reverse) const
+std::optional<std::pair<size_t, size_t>> fuzzy_searcher_impl::find_in(string_view const Haystack, bool const Reverse) const
 {
 	const auto Result = find_in_uncorrected(Haystack, Reverse);
 	if (!Result)
@@ -358,16 +356,12 @@ std::optional<std::pair<size_t, size_t>> fuzzy_searcher<Searcher>::find_in(strin
 	return Result;
 }
 
-template <typename Searcher>
-std::optional<std::pair<size_t, size_t>> fuzzy_searcher<Searcher>::find_in_uncorrected(string_view Haystack, bool Reverse) const
+std::optional<std::pair<size_t, size_t>> fuzzy_searcher_impl::find_in_uncorrected(string_view Haystack, bool Reverse) const
 {
 	normalize_for_search(Haystack, m_HayStack, m_Intermediate, m_Types);
-	return m_Searcher.find_in(m_HayStack, Reverse);
+	return m_Searcher->find_in(m_HayStack, Reverse);
 }
 
-template class fuzzy_searcher<exact_searcher>;
-template class fuzzy_searcher<icase_searcher>;
-
 #ifdef ENABLE_TESTS
 
 #include "testing.hpp"
diff --git a/far/string_utils.hpp b/far/string_utils.hpp
index 381b5e0f8..92a254d9a 100644
--- a/far/string_utils.hpp
+++ b/far/string_utils.hpp
@@ -180,20 +180,37 @@ private:
 	exact_searcher m_Searcher;
 };
 
-template <typename Searcher>
-class fuzzy_searcher final: public i_searcher
+class fuzzy_searcher_impl: public i_searcher
 {
-public:
-	NONCOPYABLE(fuzzy_searcher);
-	explicit fuzzy_searcher(string_view Needle, bool CanReverse = true);
+protected:
+	explicit fuzzy_searcher_impl(i_searcher const* Searcher);
 	std::optional<std::pair<size_t, size_t>> find_in(string_view Haystack, bool Reverse = {}) const override;
+	string needle(string_view Needle) const;
 
 private:
-	std::optional<std::pair<size_t, size_t>> find_in_uncorrected(string_view Haystack, bool Reverse = {}) const;
+	std::optional<std::pair<size_t, size_t>> find_in_uncorrected(string_view Haystack, bool Reverse) const;
 
 	mutable string m_HayStack, m_Intermediate;
 	mutable std::vector<WORD> m_Types;
 
+	i_searcher const* m_Searcher;
+};
+
+template <typename Searcher>
+class fuzzy_searcher final: public fuzzy_searcher_impl
+{
+public:
+	NONCOPYABLE(fuzzy_searcher);
+
+	explicit fuzzy_searcher(string_view const Needle, bool const CanReverse = true):
+		fuzzy_searcher_impl(&m_Searcher),
+		m_Searcher(needle(Needle), CanReverse)
+	{
+	}
+
+	using fuzzy_searcher_impl::find_in;
+
+private:
 	Searcher m_Searcher;
 };
 
```
</details>
?

Reply:
Good point. No reason duplicating the entire `find_in` implementation. I did it slightly differently though, because:
1. Giving the base class constructor a pointer to a field of the derived class, while technically allowed, is a bad form.
2. Improved encapsulation of `fuzzy_searcher` inner workings.
3. Avoided redundant copying of the `Needle` while constructing `fuzzy_searcher`.

See if you like it.

Reply:
Looks good, thank you.